Reference implementation for proposed Multi-token spec (NEP-245)

uncle-T0ny opened the pull request:
This PR is based on https://github.com/near/near-sdk-rs/pull/776
with the updated codebase and the improvements.

The improvements list (by commits): 
1) MT token: update to the latest codebase
- apply all the bugfixes in other standards that affect multi-token
- update the MT by using a new way for XCC and use the "abi" feature 

2) MT token: fix mt_is_approved, make method viewable
- in mt_is_approved view method can't use predecessor_account_id, added owner_id as a param 

3) MT token: fix the approval invalidation, reduce the approved amount
- fixed an approve invalidation bug, also decrease the approved amount after the transfer

4) MT token: added storage management, fix method types, refactor e2e tests
- implement Storage Management for MT token
- in some methods changed u128 to the wrapped version U128 for compatibility with the JS
- improved e2e tests and added more cases

frol reviewed (commented):
This is a massive PR, so I will review it in chunks. Please, bear with me

uncle-T0ny commented on examples/multi-token/test-approval-receiver/src/lib.rs:
```diff
@@ -0,0 +1,43 @@
+        env::log_str(
+            format!("Tokens: {:?} Owner: {}, approval_ids: {:?}", tokens, owner_id, approval_ids)
+                .as_str(),
+        );
```
fixed

uncle-T0ny commented on examples/multi-token/test-contract-defi/src/lib.rs:
```diff
@@ -0,0 +1,98 @@
+impl DeFi {
+    #[init]
+    pub fn new(multi_token_account_id: AccountId) -> Self {
+        require!(!env::state_exists(), "Already initialized");
```
fixed

uncle-T0ny commented on near-contract-standards/src/multi_token/approval/approval_impl.rs:
```diff
@@ -0,0 +1,178 @@
+
+            // Get the balance to check if user has enough tokens
+            let approver_balance =
+                self.balances_per_token.get(token_id).unwrap().get(&approver_id).unwrap_or(0);
```
fixed

uncle-T0ny commented on examples/multi-token/test-contract-defi/src/lib.rs:
```diff
@@ -0,0 +1,98 @@
+        token_ids: Vec<TokenId>,
+        amounts: Vec<U128>,
+        msg: String,
+    ) -> PromiseOrValue<Vec<U128>> {
```
From the [NEP-245](https://nomicon.io/Standards/Tokens/MultiToken/Core) 
```
 Take some action after receiving a multi token

 Requirements:
 * Contract MUST restrict calls to this function to a set of whitelisted  
   contracts
 * Contract MUST panic if `token_ids` length does not equals `amounts`
   length
 * Contract MUST panic if `previous_owner_ids` length does not equals `token_ids`
  length
 ```
But, you provided a nice case, I think It would be nice to add more test cases.

uncle-T0ny replied:
from other side returning Vec<(TokenId, U128)> not safe because a predecessor can change the token

frol replied:
> from other side returning Vec<(TokenId, U128)> not safe because a predecessor can change the token

What do you mean?

Well, thinking about it a bit more, I feel that the list of amounts could be cheaper/easier to implement, so maybe it is not worth complicating things

kenobijon replied:
I kept parameters as-is based on the comments above. Agree with @frol that prioritizing simplicity makes sense here

frol commented on examples/multi-token/mt/src/lib.rs:
```diff
@@ -0,0 +1,453 @@
+
+    #[init]
+    pub fn new(owner_id: AccountId, metadata: MtContractMetadata) -> Self {
+        require!(!env::state_exists(), "Already initialized");
```
Since [near-sdk-rs v3.0.1](https://github.com/near/near-sdk-rs/blob/b8fc60809b907543d909ee75fcc1bd7b68cbe2fd/CHANGELOG.md#301---2021-03-25):

> BREAKING `#[init]` now checks that the state is not initialized. This is expected behavior. To ignore state check you can call `#[init(ignore_state)]`

Thus, you don't need this check:

```suggestion
```

kenobijon replied:
removed this

frol reviewed (commented):
Thanks for addressing the previous review comments!

uncle-T0ny commented:
Update: 
Added extension to have a possibility to use Royalties and Payouts interface, like we have in NFT contract 
[MT: added token holders extension](https://github.com/near/near-sdk-rs/pull/950/commits/189de1b26c6496767a4a7ac3b491170a86641abd)

Kirill-K-1 commented on near-contract-standards/src/multi_token/approval/approval_impl.rs:
```diff
@@ -0,0 +1,178 @@
+        for (idx, (token_id, amount)) in token_ids.iter().zip(amounts).enumerate() {
+            let by_owner = by_token.get(token_id).unwrap_or_default();
+
+            let grantee_to_approval = match by_owner.get(&owner_id) {
+                Some(grantee_to_approval) => grantee_to_approval,
+                None => return false,
+            };
+
+            let approval = match grantee_to_approval.get(&approved_account_id) {
+                Some(approval) => approval,
+                None => return false,
+            };
+
+            if !approval.amount.eq(&amount.into()) {
+                return false;
+            }
```
```suggestion
            let approval = match by_owner
                .get(&owner_id)
                .and_then(|grantee_to_approval| grantee_to_approval.get(&approved_account_id))
            {
                Some(approval) if approval.amount.eq(&amount.into()) => approval,
                _ => return false,
            };
```

kenobijon replied:
Made this change

Kirill-K-1 commented on near-contract-standards/src/multi_token/token.rs:
```diff
@@ -0,0 +1,36 @@
+
+// Represents a temporary record of an Approval
+// that was removed from the ApprovalContainer but may be restored in case of rollback in XCC.
+// Values are (owner_id, approval_id, amount)
```
I think this tuple description should be changed, what do you think?

kenobijon replied:
Updated the description

Kirill-K-1 commented on near-contract-standards/src/multi_token/core/core_impl.rs:
```diff
@@ -0,0 +1,803 @@
+    pub extra_storage_in_bytes_per_emission: StorageUsage,
+
+    /// Owner of each token
+    pub owner_by_id: TreeMap<TokenId, AccountId>,
```
What is the reason of using TreeMap here?

kenobijon replied:
@Kirill-K-1 What do you propose is best data structure? Would you be fine with `UnorderedMap` so it can be iterated through?

Kirill-K-1 commented on near-contract-standards/src/multi_token/enumeration/enumeration_impl.rs:
```diff
@@ -0,0 +1,66 @@
+            self.owner_by_id.len() as u128 >= start_index,
+            "Out of bounds, please use a smaller from_index."
+        );
+        let limit = limit.map(|v| v as usize).unwrap_or(usize::MAX);
```
How about to use
```
let limit = limit.unwrap_or(u64::MAX);
```

kenobijon replied:
Fixed

uint commented:
@uncle-T0ny I'm looking through old PRs. Is this one something you'd like to keep working on and finalize? I understand if not - it looks like it moved slowly before.
